### DISCUSSION TITLE
issue adding build to group via spaceship

### DISCUSSION THREAD

**Discussion by jamesdixon** [fastlane/fastlane#19107] [Q&A]
Hi, according to https://github.com/fastlane/fastlane/issues/12984, it's possible to attach a group to a build. Is there a modern, 2021 version, of the answer provided by @joshdholtz?

I tried the following but it hangs with a timeout error when trying to grab the build:

```
    fastlane_require 'spaceship'
 
    Spaceship::Tunes.login
    Spaceship::Tunes.select_team

    app = Spaceship::ConnectAPI::App.find(ENV['APP_BUNDLE_ID'])
  
    groups = app.get_beta_groups(filter: {name: "Beta"})
    group = groups.first

    # hangs here
    build = Spaceship::TestFlight::Build.latest(app_id: app.id, platform: 'ios')
    
   # never makes it here
    build.add_group!(group)
```

Here's my Fastlane env:

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.15.7                                     |
| Ruby                        | 2.6.6                                       |
| Bundler?                    | true                                        |
| Git                         | git version 2.24.3 (Apple Git-128)          |
| Installation Source         | ~/.rvm/gems/ruby-2.6.6/bin/fastlane         |
| Host                        | Mac OS X 10.15.7 (19H1030)                  |
| Ruby Lib Dir                | ~/.rvm/rubies/ruby-2.6.6/lib                |
| OpenSSL Version             | OpenSSL 1.1.1h  22 Sep 2020                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 12.3                                        |
| Swift Version               | 5.3.2                                       |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
desc "Deploy a new version to internal users through TestFlight and Google Play"
lane :build do |options|
  begin
    self.runner.current_platform = :ios
    build

    self.runner.current_platform = :android
    build
  rescue => ex
    UI.error(ex)
  end
end

platform :android do
  desc "Deploy a new version to the Google Play internal track"
  lane :build do
    sign_and_build

    sentry_upload_proguard(
      org_slug: ENV['SENTRY_ORG_SLUG'],
      project_slug: ENV['SENTRY_PROJECT_SLUG'],
      android_manifest_path: "#{ENV['FCI_BUILD_DIR']}/build/app/intermediates/merged_manifests/release/AndroidManifest.xml",
      mapping_path: 'build/app/outputs/mapping/release/',
    )
    
    upload_to_browserstack(
      path: "build/app/outputs/bundle/release/", 
      file: "app-release.aab"
    )

    upload_to_play_store(
      track: ENV['GOOGLE_PLAY_TRACK'],
      version_code: get_build_version,
      skip_upload_screenshots: true,
      skip_upload_images: true,
      aab: "build/app/outputs/bundle/release/app-release.aab"
    )
  end

  desc "Sign and build the app for Android"
  private_lane :sign_and_build do   
    build_flutter_app(build_type: 'appbundle', version_code: get_build_version)
  end
end

platform :ios do
  KEYCHAIN_NAME = nil

  lane :build do
    # update bundle id based on environment
    # sh("git grep -l '#{ENV['APP_BUNDLE_ID']}' ios/ | xargs sed -i '' 's/#{ENV['APP_BUNDLE_ID']}/#{ENV['IOS_BUNDLE_ID']}/g'")

    create_keychain_for_ci

    # generate an api key from app store connect
    # api_key = app_store_connect_api_key(duration: 1200, is_key_content_base64: true)
    
    # sign_and_build(api_key: api_key)

    # sentry_upload_dsym(
    #   org_slug: ENV['SENTRY_ORG_SLUG'],
    #   project_slug: ENV['SENTRY_PROJECT_SLUG'],
    # )

    # upload_to_browserstack(file: "Runner.ipa")

    # upload_to_testflight(
    #   api_key: api_key,
    #   expire_previous_builds: true,
    #   notify_external_testers: false,
    #   skip_waiting_for_build_processing: true,
    #   skip_submission: true
    # )

    fastlane_require 'spaceship'
 
    Spaceship::Tunes.login
    Spaceship::Tunes.select_team

    app = Spaceship::ConnectAPI::App.find(ENV['APP_BUNDLE_ID'])
  
    groups = app.get_beta_groups(filter: {name: "Beta"})
    group = groups.first
  
    p "fetching build"  
    build = Spaceship::TestFlight::Build.all
    p build
    # build.add_group!(group)

    delete_keychain_from_ci
  end

  private_lane :sign_and_build do |values|
    sync_code_signing(
      api_key: values[:api_key],
      type: "appstore", 
      readonly: is_ci,
      keychain_name: KEYCHAIN_NAME,
      keychain_password: KEYCHAIN_NAME
    )

    version_code = get_build_version

    increment_build_number(
      build_number: version_code,
      xcodeproj: "ios/Runner.xcodeproj"
    )

    build_flutter_app(build_type: 'ios', version_code: version_code)
    
    # Build an IPA for upload to TestFlight.
    build_ipa
  end

  desc "Exports an IPA for upload. Currently can't be done by Flutter CLI."
  private_lane :build_ipa do
    build_ios_app(
      workspace: "ios/Runner.xcworkspace", 
      scheme: "Runner",
    )
  end

  private_lane :create_keychain_for_ci do
    if is_ci?
      KEYCHAIN_NAME = "temp.keychain".freeze

      create_keychain(
        name: KEYCHAIN_NAME,
        password: KEYCHAIN_NAME,
        default_keychain: true,
        unlock: true,
        timeout: 3600,
      )
    end
  end

  private_lane :delete_keychain_from_ci do
    delete_keychain(
      name: KEYCHAIN_NAME
    ) unless KEYCHAIN_NAME.nil?
  end
end

# Public Lanes

desc "Creates a release in Sentry"
lane :create_sentry_release do
  sentry_set_commits(
    version: get_full_version,
    app_identifier: ENV['APP_BUNDLE_ID'],
    auto: true, # enable completely automated commit management
    clear: false # clear all current commits from the release
  )

  sentry_create_release(
    org_slug: ENV['SENTRY_ORG_SLUG'],
    project_slug: ENV['SENTRY_PROJECT_SLUG'],
    version: get_full_version,
    app_identifier: ENV['APP_BUNDLE_ID'],
    finalize: true
  )
end

desc "Increments the build version by 1."
lane :bump_build_version do
  # commits the build version bump to the destination branch
  commit_remote_build_version_bump

  # bumps local version code so deployments get the correct value
  bump_local_build_version
end

desc "Commits the build version bump to the destination branch."
lane :commit_remote_build_version_bump do |options|
  files = [
    'pubspec.yaml', 
    'ios/Runner/Info.plist', 
    'ios/Runner.xcodeproj/project.pbxproj'
  ]
  
  destination_branch = ENV["BUILD_BUMP_DESTINATION_BRANCH"]

  sh("git checkout -b #{destination_branch}")

  Dir.chdir('..')  do
    sh("cider bump build")
  end

  git_add(path: files)
  git_commit(path: files, message: "build version bump [skip ci]")
  add_git_tag(tag: get_full_version)
  push_to_git_remote()
end

desc "Bumps the version code of the incoming branch."
lane :bump_local_build_version do |options|
  incoming_branch = git_branch

  sh("git checkout #{incoming_branch}")

  Dir.chdir('..')  do
    sh("cider bump build")
  end
end

# Private Lanes

desc "Add a badge to the app icon for visibility."
private_lane :add_badge_to_icon do |options| 
  version_code = options[:version_code]
  raise "Missing version code" if version_code.nil?

  add_badge(dark: true, alpha: true, shield: "Version-#{version_code}")
end

desc "Build a Flutter app through the CLI."
private_lane :build_flutter_app do |options|
  build_type = options[:build_type]
  raise "Missing build type" if build_type.nil?

  version_code = options[:version_code]
  raise "Missing version code" if version_code.nil?

  flutter_command = is_ci? ? "flutter" : "fvm flutter"

  Dir.chdir ".." do
    begin 
      system("#{flutter_command} packages get")
      system("#{flutter_command} clean")
      system("#{flutter_command} build #{build_type} --build-number=#{version_code}")
    rescue => ex
      UI.error ex
      UI.user_error!("There was an error during the build process.")
    end
  end
end

desc "Gets the full version from pubspec.yaml."
private_lane :get_full_version do
  begin
    pubspec = YAML.load_file('../pubspec.yaml')
  rescue
    raise 'Read pubspec.yaml failed'
  end
  
  pubspec['version']
end

desc "Gets the build version from pubspec.yaml."
private_lane :get_build_version do
  begin
    pubspec = YAML.load_file('../pubspec.yaml')
  rescue
    raise 'Read pubspec.yaml failed'
  end
  version = pubspec['version']
  
  unless version.include?('+')
    raise 'Build version indicator (+) not found in pubspec.yml'
  end

  version_sections = version.split('+')
  version_name = version_sections[0]
  version_code = version_sections[1]
  
  version_code
end

private_lane :upload_to_browserstack do |values|
  # fix for weird permissions error on CI
  # https://github.com/browserstack/browserstack-fastlane-plugin/issues/13
  system("chmod 0600 /Users/builder/.netrc")
  
  channel = values[:channel] || "internal"
  
  path = values[:path]
  file = values[:file]
  file_extension = File.extname(file)
  
  original_file_path = "#{path}#{file}"
  versioned_file_path = "#{path}spectora-x-#{channel}-#{get_full_version}#{file_extension}"
  
  system("cp ../#{original_file_path} ../#{versioned_file_path}")

  begin
    upload_to_browserstack_app_live(
      browserstack_username: ENV["BROWSERSTACK_USERNAME"],
      browserstack_access_key: ENV["BROWSERSTACK_ACCESS_KEY"],
      file_path: versioned_file_path
    )
  rescue => e
    p 'catching errors from browserstack'
    p e
  end
end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
# Android
package_name(ENV["ANDROID_BUNDLE_ID"])
json_key_file("./fastlane/google-play-upload-key.json")

# iOS
app_identifier(ENV["IOS_BUNDLE_ID"])
apple_id(ENV["FASTLANE_USER"])

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.187.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                       | Version | Update-Status      |
| ---------------------------- | ------- | ------------------ |
| fastlane-plugin-browserstack | 0.3.2   | ✅ Up-To-Date       |
| fastlane-plugin-sentry       | 1.8.1   | 🚫 Update available |

<details><summary><b>Loaded gems</b></summary>

| Gem                             | Version      |
| ------------------------------- | ------------ |
| did_you_mean                    | 1.3.0        |
| executable-hooks                | 1.6.1        |
| bundler-unload                  | 1.0.2        |
| rubygems-bundler                | 1.4.5        |
| bundler                         | 1.17.3       |
| etc                             | 1.0.1        |
| forwardable                     | 1.2.0        |
| fileutils                       | 1.1.0        |
| rake                            | 13.0.6       |
| CFPropertyList                  | 3.0.3        |
| concurrent-ruby                 | 1.1.7        |
| i18n                            | 1.8.5        |
| minitest                        | 5.14.2       |
| thread_safe                     | 0.3.6        |
| tzinfo                          | 1.2.7        |
| activesupport                   | 5.2.4.4      |
| public_suffix                   | 4.0.6        |
| addressable                     | 2.8.0        |
| httpclient                      | 2.8.3        |
| json                            | 2.5.1        |
| algoliasearch                   | 1.27.4       |
| artifactory                     | 3.0.15       |
| atomos                          | 0.1.3        |
| aws-eventstream                 | 1.1.1        |
| aws-partitions                  | 1.478.0      |
| aws-sigv4                       | 1.2.4        |
| jmespath                        | 1.4.0        |
| aws-sdk-core                    | 3.117.0      |
| aws-sdk-kms                     | 1.44.0       |
| aws-sdk-s3                      | 1.96.1       |
| babosa                          | 1.0.4        |
| claide                          | 1.0.3        |
| fuzzy_match                     | 2.0.4        |
| nap                             | 1.1.0        |
| netrc                           | 0.11.0       |
| ffi                             | 1.13.1       |
| ethon                           | 0.12.0       |
| typhoeus                        | 1.4.0        |
| cocoapods-core                  | 1.10.0.rc.1  |
| cocoapods-deintegrate           | 1.0.4        |
| cocoapods-downloader            | 1.4.0        |
| cocoapods-plugins               | 1.0.0        |
| cocoapods-search                | 1.0.0        |
| cocoapods-trunk                 | 1.5.0        |
| cocoapods-try                   | 1.2.0        |
| colored2                        | 3.1.2        |
| escape                          | 0.0.4        |
| fourflusher                     | 2.3.1        |
| gh_inspector                    | 1.1.3        |
| molinillo                       | 0.6.6        |
| ruby-macho                      | 1.4.0        |
| nanaimo                         | 0.3.0        |
| rexml                           | 3.2.5        |
| xcodeproj                       | 1.20.0       |
| cocoapods                       | 1.10.0.rc.1  |
| colored                         | 1.2          |
| highline                        | 2.0.3        |
| commander                       | 4.6.0        |
| declarative                     | 0.0.20       |
| digest-crc                      | 0.6.3        |
| unf_ext                         | 0.0.7.7      |
| unf                             | 0.1.4        |
| domain_name                     | 0.5.20190701 |
| dotenv                          | 2.7.6        |
| emoji_regex                     | 3.2.2        |
| excon                           | 0.84.0       |
| faraday-em_http                 | 1.0.0        |
| faraday-em_synchrony            | 1.0.0        |
| faraday-excon                   | 1.1.0        |
| faraday-httpclient              | 1.0.1        |
| faraday-net_http                | 1.0.1        |
| faraday-net_http_persistent     | 1.2.0        |
| faraday-patron                  | 1.0.0        |
| multipart-post                  | 2.0.0        |
| ruby2_keywords                  | 0.0.4        |
| faraday                         | 1.5.1        |
| http-cookie                     | 1.0.4        |
| faraday-cookie_jar              | 0.0.7        |
| faraday_middleware              | 1.0.0        |
| fastimage                       | 2.2.4        |
| jwt                             | 2.2.3        |
| memoist                         | 0.16.2       |
| multi_json                      | 1.15.0       |
| os                              | 1.1.1        |
| signet                          | 0.15.0       |
| googleauth                      | 0.16.2       |
| mini_mime                       | 1.1.0        |
| trailblazer-option              | 0.1.1        |
| uber                            | 0.1.0        |
| representable                   | 3.1.1        |
| retriable                       | 3.1.2        |
| webrick                         | 1.7.0        |
| google-apis-core                | 0.4.0        |
| google-apis-androidpublisher_v3 | 0.8.0        |
| google-apis-playcustomapp_v1    | 0.5.0        |
| google-apis-iamcredentials_v1   | 0.6.0        |
| google-apis-storage_v1          | 0.6.0        |
| google-cloud-env                | 1.5.0        |
| google-cloud-errors             | 1.1.0        |
| google-cloud-core               | 1.6.0        |
| google-cloud-storage            | 1.34.1       |
| mini_magick                     | 4.11.0       |
| naturally                       | 2.2.1        |
| plist                           | 3.6.0        |
| rubyzip                         | 2.3.2        |
| security                        | 0.1.3        |
| simctl                          | 1.6.8        |
| terminal-notifier               | 2.0.0        |
| unicode-display_width           | 1.7.0        |
| terminal-table                  | 1.8.0        |
| tty-screen                      | 0.8.1        |
| tty-cursor                      | 0.7.1        |
| tty-spinner                     | 0.9.3        |
| word_wrap                       | 1.0.0        |
| rouge                           | 2.0.7        |
| xcpretty                        | 0.3.0        |
| xcpretty-travis-formatter       | 1.0.1        |
| http-accept                     | 1.7.0        |
| mime-types-data                 | 3.2021.0225  |
| mime-types                      | 3.3.1        |
| rest-client                     | 2.1.0        |
| fastlane-plugin-browserstack    | 0.3.2        |
| fastlane-plugin-sentry          | 1.8.1        |
</details>

*generated on:* **2021-07-15**
</details>

**Comment by jamesdixon** [comment]
@KrauseFx do you have any suggestions? Thank you 🙇‍♂️

**Comment by lucgrabowski** [comment]
Hi, it should be possible to do it with ConnectAPI using those methods:
https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/testflight/testflight.rb#L176-L192
https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/build.rb#L149-L154
You could take a look at `pilot`'s code to see how it's used:
https://github.com/fastlane/fastlane/blob/master/pilot/lib/pilot/build_manager.rb#L396-L428
Or you could use `pilot distribute` to distribute the build to those groups.

**Comment by deiv061** [comment]
@jamesdixon  I'm experiencing the same problem, Did you find a solution 👍
